### PR TITLE
Refactor the Cloudflare client into its own library

### PR DESCRIPTION
**Pull request by coreyjonoliver**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,4 +9,18 @@ env:
 before_script:
   - jdk_switcher use $JDK
 
-script: sbt ++$TRAVIS_SCALA_VERSION clean stack/clean 'test-only -- timefactor 10' 'stack/test-only -- timefactor 10'
+script: sbt ++$TRAVIS_SCALA_VERSION test
```
Comment (bpholt) [inline .travis.yml]: I had a lot of problems with tests timing out and resulting in false negatives if the timefactors aren't set. I think we should use it here too.

File: .travis.yml [hunk 2]
```diff
@@ -9,4 +9,18 @@ env:
...
+  - tar -xzOf $HOME/secrets.tgz credentials > $HOME/.bintray/.credentials
+  - tar -xzOf $HOME/secrets.tgz release-key > $HOME/.ssh/id_rsa
+  - chmod 0600 $HOME/.ssh/id_rsa
+
+deploy:
+  - provider: script
+    script: .travis/release.sh
+    on:
+      branch: master
+      scala: 2.12.1
```
Comment (bpholt) [inline .travis.yml]: Please make sure there is a newline at the end of files.

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,7 @@
 .idea/
+.idea_modules/
 target/
 project/project/
+logs/
+.DS_Store
+*.orig
```
Comment (bpholt) [inline .gitignore]: Please make sure there is a newline at the end of files.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala [hunk 4]
```diff
@@ -13,10 +13,10 @@ import scala.language.implicitConversions
 
 class DnsRecordClient(executor: CloudflareApiExecutor)(implicit val ec: ExecutionContext) {
 
-  import Implicits._
+  import com.dwolla.cloudflare.domain.model.Implicits._
 
-  protected lazy val logger: Logger = LoggerFactory.getLogger("LambdaLogger")
-  protected implicit val formats: Formats = DefaultFormats ++ org.json4s.ext.JodaTimeSerializers.all
+  protected lazy val logger: Logger = LoggerFactory.getLogger(classOf[DnsRecordClient])
```
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: If this is going to be used in AWS Lambdas, it should use the `"LambdaLogger"` logger, so that CloudWatch logging doesn't split the log stream into multiple log events using newlines.

File: build.sbt [hunk 5]
```diff
@@ -1,82 +1,52 @@
...
-  organization := "Dwolla",
-  homepage := Option(url("https://stash.dwolla.net/projects/OPS/repos/cloudflare-public-hostname-lambda")),
+lazy val buildSettings = Seq(
+  organization := "com.dwolla",
+  name := "cloudformation",
+  homepage := Some(url("https://github.com/Dwolla/scala-cloudflare")),
+  description := "Scala library for the Cloudflare v4 API",
+  licenses += ("MIT", url("http://opensource.org/licenses/MIT")),
+  releaseVersionBump := sbtrelease.Version.Bump.Minor,
+  releaseProcess := {
```
Comment (bpholt) [inline build.sbt]: This release process excludes the publish step. I think since we're not building for multiple Scala versions we can remove this customization and rely on the default `releaseProcess`.

File: build.sbt [hunk 6]
```diff
@@ -1,82 +1,52 @@
...
-}
-test in assembly := {}
-
-lazy val slf4jEarlyInit = sbt.Tests.Setup(cl ⇒
-  cl.loadClass("org.slf4j.LoggerFactory")
-    .getMethod("getLogger", cl.loadClass("java.lang.String"))
-    .invoke(null, "ROOT")
+lazy val bintraySettings = Seq(
+  bintrayVcsUrl := homepage.value.map(_.toString),
+  publishMavenStyle := false,
```
Comment (bpholt) [inline build.sbt]: Does this need to be true if we're publishing to the `maven` repo?

File: build.sbt [hunk 7]
```diff
@@ -1,82 +1,52 @@
...
+  bintrayRepository := "maven",
+  bintrayOrganization := Option("dwolla"),
+  pomIncludeRepository := { _ ⇒ false }
 )
 
-testOptions += slf4jEarlyInit
+lazy val root = (project in file("."))
+  .settings(buildSettings ++ bintraySettings: _*)
```
Comment (bpholt) [inline build.sbt]: Please make sure there is a newline at the end of files.

File: project/plugins.sbt [hunk 8]
```diff
@@ -1,8 +1,3 @@
 logLevel := Level.Warn
-addSbtPlugin("com.dwolla.sbt" % "sbt-s3-publisher" % "1.0.0")
-addSbtPlugin("com.dwolla.sbt" % "sbt-cloudformation-stack" % "1.2.0")
-
-resolvers ++= Seq(
-  Resolver.bintrayIvyRepo("dwolla", "sbt-plugins"),
-  Resolver.bintrayIvyRepo("dwolla", "maven")
-)
+addSbtPlugin("me.lessis" % "bintray-sbt" % "0.3.0")
+addSbtPlugin("com.github.gseitz" % "sbt-release" % "1.0.4")
```
Comment (bpholt) [inline project/plugins.sbt]: Please make sure there is a newline at the end of files.

File: src/main/scala/com/dwolla/cloudflare/CloudflareApiExecutor.scala [hunk 9]
```diff
@@ -2,21 +2,19 @@ package com.dwolla.cloudflare
...
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.Try
 import scala.language.implicitConversions
 
 class CloudflareApiExecutor(authorization: CloudflareAuthorization)(implicit val ec: ExecutionContext) extends Closeable {
   lazy val httpClient: CloseableHttpClient = HttpClients.createDefault()
 
   def fetch[T](request: HttpRequestBase)(f: HttpResponse ⇒ T): Future[T] = {
-    val variableToForceImplicitConversion = Future {
+    Future.fromTry {
```
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/CloudflareApiExecutor.scala]: +1

File: src/main/scala/com/dwolla/cloudflare/common/JsonEntity.scala [hunk 10]
```diff
@@ -22,4 +22,4 @@ object JsonEntity {
   implicit def anyRefToJsonEntity[T <: AnyRef, S <: JsonWritable](x: T)(implicit formats: Formats, ev: T ⇒ S): JsonEntity[S] = JsonEntity(ev(x))
 }
 
-trait JsonWritable extends AnyRef
+trait JsonWritable extends AnyRef
```
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/common/JsonEntity.scala]: Please make sure there is a newline at the end of files.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala [hunk 11]
```diff
@@ -84,16 +82,16 @@ class DnsRecordClient(executor: CloudflareApiExecutor)(implicit val ec: Executio
...
+
   implicit def httpResponseToJsonInput(httpResponse: HttpResponse): MonadicJValue = parseJson(httpResponse.getEntity.getContent)
 }
 
 case class MultipleCloudflareRecordsExistForDomainNameException(domainName: String, records: Set[DnsRecordDTO]) extends RuntimeException(
   s"""Multiple DNS records exist for domain name $domainName:
      |
      | - ${records.mkString("\n - ")}
-     |
-     |This resource refuses to process multiple records because the intention is not clear.
```
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: Why is this clarification being removed?

Comment (coreyjonoliver) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: I was under the impression this line was AWS Lambda specific. I can add it back if that's not the case.

Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: Well, the next line mentions the Lambda, but it could be changed to reflect the library being the thing that needs to change instead of the Lambda function itself. AFAICT the logic of the library still enforces the no-multiple-records rule.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala (hunk 4 above)
Comment (coreyjonoliver) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: I noticed the `logger` variable isn't even used in the portion I pulled out. You mind if I just remove the reference?

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala (hunk 11 above)
Comment (coreyjonoliver) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: I can rephrase it to make sense in the context of a library.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala (hunk 4 above)
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: Yep, good call.

File: .travis.yml (hunk 1 above)
Comment (coreyjonoliver) [inline .travis.yml]: I'll add it.

File: build.sbt (hunk 5 above)
Comment (coreyjonoliver) [inline build.sbt]: I'll remove `releaseProcess`.

File: build.sbt (hunk 6 above)
Comment (coreyjonoliver) [inline build.sbt]: I just used the same settings specified in https://github.com/Dwolla/scala-aws-utils/blob/master/build.sbt#L41. I have no problem changing it if required though.

Comment (bpholt) [inline build.sbt]: Fair enough… it's probably fine since AFAIK it's working there. ¯\_(ツ)_/¯

File: .travis.yml (hunk 2 above)
Comment (coreyjonoliver) [inline .travis.yml]: Addressed.

File: .gitignore (hunk 3 above)
Comment (coreyjonoliver) [inline .gitignore]: Addressed.

File: build.sbt (hunk 5 above)
Comment (coreyjonoliver) [inline build.sbt]: Addressed.

File: build.sbt (hunk 7 above)
Comment (coreyjonoliver) [inline build.sbt]: Addressed.

File: project/plugins.sbt (hunk 8 above)
Comment (coreyjonoliver) [inline project/plugins.sbt]: Addressed.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala (hunk 11 above)
Comment (coreyjonoliver) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: Addressed.

File: src/main/scala/com/dwolla/cloudflare/common/JsonEntity.scala (hunk 10 above)
Comment (coreyjonoliver) [inline src/main/scala/com/dwolla/cloudflare/common/JsonEntity.scala]: Addressed.

File: .travis.yml (hunk 1 above)
Comment (coreyjonoliver) [inline .travis.yml]: Addressed.

File: build.sbt [hunk 12]
```diff
@@ -1,82 +1,38 @@
-javacOptions ++= Seq("-source", "1.8", "-target", "1.8", "-Xlint")
-
-lazy val commonSettings = Seq(
-  organization := "Dwolla",
-  homepage := Option(url("https://stash.dwolla.net/projects/OPS/repos/cloudflare-public-hostname-lambda")),
+lazy val buildSettings = Seq(
+  organization := "com.dwolla",
+  name := "cloudformation",
```
Comment (bpholt) [inline build.sbt]: The artifact name `cloudformation` ought to be something like `cloudflare-api-client`. (I think just calling it `cloudflare` would violate [their trademark terms of service](https://www.cloudflare.com/trademark/), but `cloudflare-api-client` uses the trademark in a descriptive way and would be fine? Idk, IANAL)

File: README.md [hunk 13]
```diff
@@ -1,74 +1,7 @@
-# Cloudflare DNS Record Handler
+# scala-cloudflare
 
-[![Travis](https://img.shields.io/travis/Dwolla/cloudflare-public-hostname-lambda.svg?style=flat-square)](https://travis-ci.org/Dwolla/cloudflare-public-hostname-lambda)
-![license](https://img.shields.io/github/license/Dwolla/cloudflare-public-hostname-lambda.svg?style=flat-square)
+[![Travis](https://img.shields.io/travis/Dwolla/scala-cloudflare.svg?style=flat-square)](https://travis-ci.org/Dwolla/scala-cloudflare)
+[![Bintray](https://img.shields.io/bintray/v/dwolla/maven/cloudflare.svg?style=flat-square)](https://bintray.com/dwolla/maven/cloudflare/view)
```
Comment (bpholt) [inline README.md]: This should be changed to match the artifact name set in `build.sbt`.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala [hunk 14]
```diff
@@ -93,7 +91,7 @@ case class MultipleCloudflareRecordsExistForDomainNameException(domainName: Stri
      | - ${records.mkString("\n - ")}
      |
      |This resource refuses to process multiple records because the intention is not clear.
-     |Clean up the records manually or enhance the custom resource Lambda to handle multiple records.""".stripMargin)
+     |Clean up the records manually or update your code to handle multiple records.""".stripMargin)
```
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: This is nitpicky but it's not the code using this library that would need to be updated: it's the library itself. This exception is thrown on line 45 if an existing DNS record is looked up by name (e.g. www.dwolla.com) and multiple records are returned from Cloudflare. The library doesn't know how to handle that situation, so it throws the exception. This error message is trying to be explicit about the implementation choice that was made, i.e., to give up instead of trying to figure out which of the multiple records should be updated.

File: src/main/scala/com/dwolla/cloudflare/domain/model/DnsRecord.scala [hunk 15]
```diff
@@ -1,4 +1,4 @@
-package com.dwolla.cloudflare.model
+package com.dwolla.cloudflare.domain.model
```
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/domain/model/DnsRecord.scala]: Just curious, why did you rename this package?

File: build.sbt (hunk 12 above)
Comment (coreyjonoliver) [inline build.sbt]: Yeah. It should definitely be changed to something containing `cloudflare` --- my bad.

File: README.md (hunk 13 above)
Comment (coreyjonoliver) [inline README.md]: Will update.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala (hunk 14 above)
Comment (coreyjonoliver) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: Gotcha. Any preference on how you would have it worded?

File: src/main/scala/com/dwolla/cloudflare/domain/model/DnsRecord.scala (hunk 15 above)
Comment (coreyjonoliver) [inline src/main/scala/com/dwolla/cloudflare/domain/model/DnsRecord.scala]: It just emphasizes the classes contained in the package are a model of our domain. It's a convention I borrowed from Vaughn Vernon's book "Implementing Domain-Driven Design". See, for example, https://github.com/VaughnVernon/IDDD_Samples/tree/master/iddd_agilepm/src/main/java/com/saasovation/agilepm/domain/model.

File: build.sbt (hunk 12 above)
Comment (coreyjonoliver) [inline build.sbt]: Addressed.

File: README.md (hunk 13 above)
Comment (coreyjonoliver) [inline README.md]: Addressed.

File: src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala (hunk 14 above)
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/DnsRecordClient.scala]: Idk… “enhance this library”?

File: src/main/scala/com/dwolla/cloudflare/domain/model/DnsRecord.scala (hunk 15 above)
Comment (bpholt) [inline src/main/scala/com/dwolla/cloudflare/domain/model/DnsRecord.scala]: Ok. I guess if these classes are only used inside the library, then it won't impact the original project and makes sense to change. If we'll need to go back and update a bunch of imports in the original project, then I probably would have left it alone. Just my preference, though—I'm ok with leaving it now that the change has been made.
